coderd/httpmw/ratelimit.go

```diff
@@ -70,3 +72,70 @@ func RateLimit(count int, window time.Duration) func(http.Handler) http.Handler
 		}),
 	)
 }
+
+// RateLimitByAuthToken returns a handler that limits requests based on the
+// authentication token in the request. It uses APITokenFromRequest to extract
+// the token, with an additional fallback to the X-Api-Key header for AI provider
+// compatibility. If no token is found, it falls back to rate limiting by IP address.
+//
+// This is useful for endpoints that handle authentication internally rather
+// than via ExtractAPIKeyMW middleware, providing per-user rate limiting.
+func RateLimitByAuthToken(count int, window time.Duration) func(http.Handler) http.Handler {
+	if count <= 0 {
+		return func(handler http.Handler) http.Handler {
+			return handler
+		}
+	}
+
+	return httprate.Limit(
+		count,
+		window,
+		httprate.WithKeyFuncs(func(r *http.Request) (string, error) {
+			// Try to extract auth token for per-user rate limiting.
+			// First check standard Coder auth methods.
+			if token := APITokenFromRequest(r); token != "" {
+				return token, nil
+			}
+			// Fallback to X-Api-Key header for AI provider compatibility
+			// (e.g., Anthropic uses this header).
+			if apiKey := r.Header.Get("X-Api-Key"); apiKey != "" {
+				return strings.TrimSpace(apiKey), nil
+			}
```

Comment:
This logic already exists in `enterprise/aibridged/http.go` -> `ExtractAuthToken`.